integration_test/upgrade_test.go

```diff
@@ -14,4 +14,50 @@
 
 package integrationtest
 
-// TODO(ahmetb): implement upgrade tests (https://krew.dev/issues/233)
+import (
+	"crypto/sha256"
+	"io"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+func TestKrewUpgrade(t *testing.T) {
+	skipShort(t)
+
+	test, cleanup := NewTest(t)
+	defer cleanup()
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	test.WithIndex().
+		Krew("install", "--manifest", filepath.Join(cwd, "testdata", "konfig.yaml")).
+		RunOrFail()
+	location := test.AssertExecutableInPATH("kubectl-" + validPlugin)
+	initialHash := hashFile(t, location)
```

Comment:
I’m -1 for reading file contents. Even if you did, you don't need a hash like sha256, you can just `reflect.DeepEquals` the bytes –or even simply compare `len(bytes)`. :)

Also I don't want to rely on os.Open behavior on symlinks, though I know open() follows symlinks. 

What I would do here is a simple `readlink()` and see if the symlink paths are the same. We know that upon upgrade, they would change.

(I realize that you can make an argument here about testing the behavior; not the implementation. And comparing bytes directly is perhaps a better idea.)

Reply:
Another option is this:

```
fi1 := os.Lstat("kubectl-foo")
...
fi2 := os.Lstat("kubectl-bar")
if os.SameFile(fi1,fi2) {...
```

Reply:
I went with the readlink approach because `Lstat` may have problems when one of the files has been deleted in the meantime.